### PR TITLE
stash: split out a stash-types crate

### PR DESCRIPTION
**Pull request by danhhz**
This allows for faster iteration on mz-stash itself, as is happening in Pv2.

Conceptually, stash is only used in envd (by the catalog and by the storage controller) but it was being pulled in by all sorts of crates for the objects protos. Separate out the latter (plus StashError) into a new mz-stash-types crate that is still depended on by lots of things, but changes less often.

The various old versions of objects.proto want to be next to it, so pull them into mz-stash-types, too. In an ideal world, we'd probably move the vX_to_vX+1 upgrades also, but they depend on actual stash things (e.g. Transaction), so leave them where they are.

Finally, any given metric can only be registered once (otherwise we get a panic), but we need two stashes, so also pull out the stash Metrics into mz-stash-types. Ideally, this would have stayed where it was. A possible alternative to this would be something like putting a type-safe cache of registered metrics objects on MetricsRegistry.

Timing on my M1 laptop of bin/environmentd after invalidating stash went from 2m12s to 53s. A large part of this is likely that we've entirely eliminated clusterd, including linking.

```
touch src/stash/src/lib.rs && time ./bin/environmentd -- --nope
```

Deps before:

deps-stash-before

Deps after:

deps-stash-after

### Motivation

   * This PR refactors existing code.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/stash/src/upgrade/v28_to_v29.rs [hunk 1]
```diff
@@ -10,8 +10,8 @@
 use std::collections::BTreeSet;
 
 use crate::upgrade::MigrationAction;
-use crate::upgrade::{objects_v28, objects_v29};
 use crate::{StashError, Transaction, TypedCollection};
+use mz_stash_types::upgrade::{objects_v28, objects_v29};
```
Comment (danhhz) [inline src/stash/src/upgrade/v28_to_v29.rs:14]: Will fix up these use blocks, too

**Comment by aljoscha** [comment]
those before/after dependency graphs... 😂

File: src/stash/src/upgrade/v28_to_v29.rs [hunk 2]
```diff
@@ -74,149 +74,133 @@ pub async fn upgrade(tx: &'_ mut Transaction<'_>) -> Result<(), StashError> {
...
-            }
-            Some(objects_v28::role_id::Value::System(id)) => {
-                Some(objects_v29::role_id::Value::System(id))
-            }
-            Some(objects_v28::role_id::Value::Public(_)) => {
-                Some(objects_v29::role_id::Value::Public(Default::default()))
-            }
-        };
-        objects_v29::RoleId { value }
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/stash/src/upgrade/v28_to_v29.rs:91]: I'm not sure, it's possible I left these in by mistake and they weren't needed. Why are you changing all of these `From` impls into normal functions?

[the remark this answers is not in the document]
Comment (jkosh44) [reply src/stash/src/upgrade/v28_to_v29.rs:91]: Ah ok, I looked through and I don't see this used anywhere. I probably just added it out of instinct after adding the database id function above.

[the remark this answers is not in the document]
Comment (danhhz) [reply src/stash/src/upgrade/v28_to_v29.rs:91]: Think the other ones are the same story? (There are a few)

**Comment by danhhz** [comment]
TFTR!

**Comment by danhhz** [comment]
Oh lolol cargo check on just mz-stash is even more dramatic (because the protos all got moved out)

Before

```
$ touch src/stash/src/lib.rs && time cargo check -p mz-stash --lib
    Checking mz-stash v0.0.0 (/Users/dan/mz/materialize/src/stash)
    Finished dev [unoptimized + debuginfo] target(s) in 35.01s

real	0m35.141s
```

After

```
$ touch src/stash/src/lib.rs && time cargo check -p mz-stash --lib
    Checking mz-stash v0.0.0 (/Users/dan/mz/materialize/src/stash)
    Finished dev [unoptimized + debuginfo] target(s) in 2.00s

real	0m2.081s
```

**Pull request merged**